### PR TITLE
Improve channel password handling

### LINKED ISSUE #3856: [Feature Request] UI indication for password-protected channels
At the moment there seems to be no indication in Mumble's UI for password-protected channels. Thus making it impossible to tell which channels are password-protected and which are not.

Not even the error message `You were denied Enter privileges in XYZ` makes it clear that the channel is password-protected (one could think one simply doesn't have the permission to enter it when instead only the PW is missing).

I think adding a lock-symbol in the channel "bar" would be nice - See this TS3 screenshot for what I mean: http://img.creativemark.co.uk/uploads/images/558/11558/largeImg.png

In terms of implementation these links might be useful:
https://stackoverflow.com/questions/12316309/how-to-set-an-image-for-a-row
https://stackoverflow.com/questions/22093510/display-image-in-qtreeview

### LINKED ISSUE #3857: [Feature Request] Improve handling of password-protected channels
When attempting to connect to a password-protected channel (without having the respective access-token set) the user gets a generic error message of type
```
You were denied Enter privileges in XYZ
```
From the error message one could assume that one doesn't have the permission to access the channel. Therefore I'd suggest to change the message to include something like `Wrong password` or `No password provided`.

Also I think that the attempt to enter a password-protected channel could optionally trigger a pop-up for entering the password for that channel. And on successful entering, the password could be stored permanently into the user's access tokens.

### PR DESCRIPTION
**Pull request by Krzmbrzl**
This is aimed to fix #3857 and to fix #3856

Changes of this PR:
- Add 2 fields to the `MumbleProto::channelState` msg indicating whether a channel is protected by ACLs denying ENTER privilege and whether the user the message is being sent to has the permissions to enter the channel
- Implement the backend to process the abovementioned fields (Mumble + Murmur)
- Added the equivalents of those message-fields as fields in the channel class inside the Mumble-client
- UI: Show lock-icon if channel is access-restricted (locked lock for the case in which the local user can't enter the respective channel and unlocked lock for the case in which the local user has the necessary privileges to enter the channel noetheless)
- UI (kinda): Changed permission denied message for ENTER or MOVE requests, if the channel has access-restrictions (the message now states that a missing password/access token might be the cause for the denial)

### CODE REVIEW

**Comment by Krzmbrzl** [comment]
I implemented #3856 and this is how it looks now:

Mumble_PWIcons_Lite

Note that there are 2 different symbols: A closed (red) lock and an opened (green) lock. The former is shown, if the channel is password-protected and you don't have a matching access token set. The latter is shown, if the channel is password-protected, but you have a matching access token configured. These locks don't actually indicate whether you can enter those channels though as they don't take other ACLs into account (clearly seen by the fact that even the SuperUser is presented with a red lock).

What do you think about this feature?

**Comment by Krzmbrzl** [comment]
Btw.: This is how it looks for the classic theme. I used the icons suggest in https://github.com/mumble-voip/mumble/pull/3872#issuecomment-547796422 for it.

Mumble_PWIcons_Classic

**Comment by Krzmbrzl** [comment]
I guess this PR is pretty much done for now. Once #3927 gets merged, I could add a popup asking for a password once the permission denied message (most likely associated with invalid password / access token) is received and then try again with that password set...

**Comment by Krzmbrzl** [comment]
This PR depends on #3933 and should be merged after it.

**Comment by Johni0702** [comment]
I strongly disagree with sending password hashes to all clients as that'll allow any client to crack them by offline brute force.
As is, a client would have to send two packets (one for setting its tokens and one for testing if they worked, latter ones afaict have a strict rate limit) per X passwords they'd like to try (limited by packet size iirc), making brute force infeasible (or at least non-trivial). An offline attack is multiple orders of magnitude faster (e.g. [this](https://security.stackexchange.com/a/18193) SO post names ~100M sha512 hashes per second, it's from 2012) making it almost trivial to brute force channel passwords (which generally won't be particularly strong by their nature).

Less relevant given the above is already a deal breaker but still worth mentioning:
- The salt seems fairly pointless given it's the same every time
- Do not use SHA512 for password hashing, it's fast by design
- Unless you know exactly what you're doing, don't roll your own crypto

Instead I'd suggest adding two bool flags (e.g. "has_lock", "is_locked") to the ChannelState message and comparing tokens on the server-side. And make sure there's no way to update them without a rate limit (probably best to just add a rate limit to the Authenticate message).

**Comment by Krzmbrzl** [comment]
Thanks for your input.

First of all, let me state that I don't see channel passwords to be such a valuable target that someone might actually go through and bruteforce them (and discussions like #3926 suggest to me that they aren't considered that valuable to others) 

That being said though, I agree that if we manage to get the features with a safer approach, that's definitely better.

> The salt seems fairly pointless given it's the same every time

As the client needs to be able to also compute the hash, I don't see a good way of making a variable salt (unless maybe randomly generated and transmitted to the client on connect. Though that of course kinda breaks the purpose of a salt, I guess). I added the salt simply to prevent someone not affiliated with Mumble's code base from using rainbow tables to crack the hash. I'm far from being an expert on these things though, so feel free to correct me in that (and anything that follows). 

> Do not use SHA512 for password hashing, it's fast by design

What would you suggest instead? 

> Unless you know exactly what you're doing, don't roll your own crypto

Does the current state if this PR fall under this? :thinking:

> Instead I'd suggest adding two bool flags (e.g. "has_lock", "is_locked") to the ChannelState message and comparing tokens on the server-side.

So every time ACL change, the server shall update all clients with new, personalized channel info? I guess that could work as well. Dunno if this could lead to too much overhead on the server-side, though I can't really imagine this getting a problem. 
I'll give this some thought, but on first glance, this definitely is a good alternative. 
Plus it has the advantage of also being able to respect ACLs other than the access tokens, making the approach much more universal :)

**Comment by Kissaki** [comment]
If possible please keep PRs as distinctive and simple as possible. This now seems to tackle multiple things in multiple commits and without an explanation in the PR description making it harder to review and get into/understand. It also makes it impossible to land partials/selected parts.

**Comment by Krzmbrzl** [comment]
> This now seems to tackle multiple things in multiple commits and without an explanation in the PR description making it harder to review and get into/understand

Yes it does 2 things. But the second one is simply changing the error message if ENTER privliges were denied in certain cases and this change is only possible because of the changes im the first one.
Therefore the only option would have been to create 3 PRs: backend, feature1, feature2
I don't think that this would contribute a lot to readability of the PR, if one starts to make multiple PRs that are dependent on one another...

The very short "description" I provided was originally intended to be expanded, but I forgot. Sorry for that (I'll do that right now).

**Comment by Krzmbrzl** [comment]
Oh and the dependency to #3933 is gone in the latest refactoring. Hopefully I'll find the time to squash the commits in this PR tomorrow to clean up the current mess xD

**Comment by Krzmbrzl** [comment]
@Kissaki I restructured my commits and from the commit messages alongside what else is written in this PR, I hope it will be relatively easy to follow the changes along.

Note also that the build fails, because https://github.com/mumble-voip/mumble-theme/pull/21 is not merged yet. once that is merged, the code in this PR should build just fine (as shown by the CI of the second last commit.

**Comment by davidebeatrici** [comment]
Is backwards compatibility maintained?

**Comment by Krzmbrzl** [comment]
I don't see how it wps break it. The whole thing is implemented by extending the protocol by two fields. A server that doesn't know it, won't populate the fields in the first place and a client that doesn't know them won't read them. Each of these cases will result in the icons not being shown, but as far as I can tell, there shouldn't be a problem other than that.

I haven't actually tested the above though. Might be worth a try. However it might take a bit until I'll be able to do so, as my beloved Linux notebook is in repair and I just don't want to fiddle around with setting up a build environment on Windows xD

File: src/ACL.cpp [hunk 1]
```diff
@@ -238,3 +238,18 @@ QString ChanACL::permName(Perm p) {
 	}
 	return QString();
 }
+
+bool ChanACL::isPassword() const {
+	// A password is an ACL that applies to a group of the form '#<something>'
+	// AND grants 'Enter'
```
Comment (Krzmbrzl) [inline src/ACL.cpp]: This comment has to be corrected in order to correspond to the actual implementation. See #3928

**Comment by Krzmbrzl** [comment]
So I did test connecting to a normal server and there was no problem doing that. And I just can't imagine that there is a problem for a client connecting to a Server running with this patch as it simply won't process the extra information the Server now provides :)

**Comment by toby63** [comment]
Sry but as I already said in https://github.com/mumble-voip/mumble-theme/issues/17#issuecomment-641387715 I think this implementation is too confusing.
If a user sees the green symbol they assume that they can enter the channel, but are asked for a password, this is inconsistent design.

I think about two solutions:
1. For channels which are access-restricted by ACL, #3057 should be applied, so making them invisible for normal users (or everyone who is not admin or group member).
2. For channels which are only access-restricted via password, use a lock-symbol to indicate that it is access-restricted.

To seperate these two scenarios better from each other, you might apply a different ACL-system.
We would have two groups:
1. Exclusive Groups (apply to scenario 1 see above): they cannot be joined via a password, only via an acl that is set by an admin or group leader or group member etc.
2. Channel Groups (aka non-exclusive groups) (apply to scenario 2 see above): they can be joined via a password and are visible to all normal users.

Maybe technological my concept is already doable, but the UI should also represent this.

**Comment by Krzmbrzl** [comment]
> If a user sees the green symbol they assume that they can enter the channel, but are asked for a password, this is inconsistent design.

Where did you get that from?
If the lock is green, the user can enter the channel without any further actions. Only when it's red, that user has to configure the access tokens (password) accordingly in order to access the channel.

**Comment by Krzmbrzl** [comment]
> For channels which are access-restricted by ACL, #3057 should be applied, so making them invisible for normal users (or everyone who is not admin or group member).

I disagree 100%. If I don't even see the channel, I would never think about setting my access tokens. This would confuse the hell out of our users

> For channels which are only access-restricted via password, use a lock-symbol to indicate that it is access-restricted.

The password is in fact also only an ACL

**Comment by toby63** [comment]
Very confusing. I don't have the time to dig into that. If you think that everyone can understand that, then let it be.

> I would never think about setting my access tokens

What should that mean?
Either you are the admin and can set access tokens or not.

> The password is in fact also only an ACL

I know that.
But a I already clarified 
>> Maybe technology-wise my concept is already (partly) implemented, but the UI should also represent this.

this is about meta-functionality, if you can follow me.
Because thats what is needed for people to understand the system correctly.

**Comment by toby63** [comment]
> > If a user sees the green symbol they assume that they can enter the channel, but are asked for a password, this is inconsistent design.
> 
> Where did you get that from?
> If the lock is green, the user can enter the channel without any further actions. Only when it's red, that user has to configure the access tokens (password) accordingly in order to access the channel.

I see, I missunderstood this whole thing then.
I thought the unlocked-icon would indicate that you have enter permission as a normal user, but would need to type in a password to join the channel.
And the locked-icon indicates that you can't join it, because you are not part of the right group and also cannot join via a password.

Nonetheless you should streamline this concept of access-tokens, passwords and acls, it's very confusing as it is now.
And to remind that once again, it is not necessarily about changes behind the scenes, but UI changes etc.

**Comment by Krzmbrzl** [comment]
> Nonetheless you should streamline this concept of access-tokens, passwords and acls, it's very confusing as it is now.

green & unlocked == you can enter
red & locked == you can't enter

what is confusing about this?

**Comment by toby63** [comment]
As I said I missunderstood the functionality, still this is not perfect.

> red & locked == you can't enter

Thats for example a potencial point of confusion.
A user might ask "is it impossible to join the channel? or is it just password-protected?".
With only one neutral lock symbol (like the golden symbol already mentioned) this is not so confusing,

> what is confusing about this?

Well this specific lock symbol concept is not perfect, but somewhat ok.
What I mean in the first place is the overall concept.
This mix of ACL, Groups, Password, Access Tokens etc.
I think it should be changed to be more clear.
#3247 is one point.
Another point is the confusing mix of password, access tokens and groups.
But I don't really know a way to solve this, without making it more complicated.
I thought about a pure password option (groups can still be used in the background, but would be deleted on password change).
But that would complicate it, because of one more option.

Anyway I also don't fully understand the whole functionality, so as of now I am no help in redesigning it.

**Comment by Krzmbrzl** [comment]
> With only one neutral lock symbol (like the golden symbol already mentioned) this is not so confusing,

One could argue that the user could ask himself the same question here. Additionally that user might be wondering if he can join the channel right now.

Besides: Whether you can't join due to not knowing the password or due to groups doesn't actually matter. The effect is the same. And once somebody gives you the password, you add it to your access tokens and then you'll see what channels it will grant access to.

**Comment by Krzmbrzl** [comment]
Oh and just to avoid misunderstandings: This PR of mine didn't introduce channel passwords or the way restricted channels are handled. I merely added an indication of it to the UI :point_up:

**Pull request merged**